src/ngraph/state/rng_state.hpp

```diff
@@ -0,0 +1,52 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <random>
+
+#include "state.hpp"
+
+namespace ngraph
+{
+    //can be based on TensorSate to cache values instead of just caching seed
+    class RNGState : public State
+    {
+    public:
+        static RNGState* create_rng_state(unsigned int seed, double probability)
+        {
+            auto rng = new RNGState(seed, probability);
+            return rng;
+        }
+
+        RNGState(unsigned int seed, double probability)
+            : State()
+            , m_generator(seed)
+            , m_distribution(probability)
+        {
+        }
+        virtual void activate() override;
+        virtual void deactivate() override;
+        virtual ~RNGState() {}
+        std::mt19937& get_generator() { return m_generator; }
```

Comment:
As far as I understand that type is just a typedef of another concrete type : `typedef mersenne_twister_engine<uint_fast32_t, 32,624,397,31,0x9908b0df,11,0xffffffff,7,0x9d2c5680,15,0xefc60000,18,1812433253> mt19937;`; Random device is another **concrete** type of a random number generator

Reply:
I think this is okay for now; multinomial is probably next, and will give us a better picture of the lay of the land.

Reply:
@diyessi I think the next step is to provide an interface to abstract random number generator and then probability objects because other backends might have different implementations and in stdlib all those are concrete types, so we need some kind of a wrapper with well defined API